### PR TITLE
Upgrade to Jersey 2.28

### PR DESCRIPTION
**Pull request by joschi**

### CODE REVIEW

File: dropwizard-jersey/src/test/java/io/dropwizard/jersey/validation/ConstraintViolationExceptionMapperTest.java [hunk 1]
```diff
@@ -323,24 +324,25 @@ public void getInvalidMatrixParamIs400() {
                 .matrixParam("bob", "").request().get();
         assertThat(response.getStatus()).isEqualTo(400);
 
-        String ret = "{\"errors\":[\"matrix param bob may not be empty\"]}";
+        String ret = "{\"errors\":[\"matrix param bob must not be empty\"]}";
         assertThat(response.readEntity(String.class)).isEqualTo(ret);
     }
 
     @Test
+    @Ignore("FIXME")
```
Comment (joschi) [inline dropwizard-jersey/src/test/java/io/dropwizard/jersey/validation/ConstraintViolationExceptionMapperTest.java]: This test is failing and the issue has to be fixed.

File: dropwizard-jersey/src/test/java/io/dropwizard/jersey/validation/ConstraintViolationExceptionMapperTest.java [hunk 2]
```diff
@@ -572,24 +574,26 @@ public void testInvalidFieldQueryParam() {
...
```
Comment (joschi) [inline dropwizard-jersey/src/test/java/io/dropwizard/jersey/validation/ConstraintViolationExceptionMapperTest.java]: This test is failing and the issue has to be fixed.

File: dropwizard-jersey/src/test/java/io/dropwizard/jersey/validation/ConstraintViolationExceptionMapperTest.java [hunk 3]
```diff
@@ -572,24 +574,26 @@ public void testInvalidFieldQueryParam() {
...
             .request()
             .get();
         assertThat(response.getStatus()).isEqualTo(400);
         assertThat(response.readEntity(String.class))
-            .containsOnlyOnce("query param length may not be null");
+            .containsOnlyOnce("query param length must not be null");
     }
 
     @Test
+    @Ignore("FIXME")
```
Comment (joschi) [inline dropwizard-jersey/src/test/java/io/dropwizard/jersey/validation/ConstraintViolationExceptionMapperTest.java]: This test is failing and the issue has to be fixed.

File: dropwizard-jersey/src/test/java/io/dropwizard/jersey/AbstractJerseyTest.java [hunk 4]
```diff
@@ -12,7 +12,7 @@
 public abstract class AbstractJerseyTest extends JerseyTest {
 
     private static final int DEFAULT_CONNECT_TIMEOUT_MS = 1000;
-    private static final int DEFAULT_READ_TIMEOUT_MS = 5000;
+    private static final int DEFAULT_READ_TIMEOUT_MS = 500000;
```
[the remark this answers is not in the document]
Comment (joschi) [reply dropwizard-jersey/src/test/java/io/dropwizard/jersey/AbstractJerseyTest.java]: Good catch, thanks!

This change was required for debugging Jersey-based without running into the 5 seconds timeout every time. It will be reverted before merging this PR. 😉

File: dropwizard-testing/src/test/java/io/dropwizard/testing/app/PersonResource.java [hunk 5]
```diff
@@ -48,7 +50,8 @@ public Person getPerson(@PathParam("name") String name) {
     @GET
     @Timed
     @Path("/index")
-    public Person getPersonWithIndex(@Min(0) @QueryParam("ind") IntParam index,
+    public Person getPersonWithIndex(@QueryParam("ind")
+                                     @NotNull(payload = Unwrapping.Skip.class) @Min(0) IntParam index,
```
Comment (usulkies) [inline dropwizard-testing/src/test/java/io/dropwizard/testing/app/PersonResource.java]: Why did you add the `@NotNull` validation?

**Comment by joschi** [comment]
@usulkies Good catch, thanks! Thankfully, it was quite simple to replace the HK2 binders with Jersey binders.

Unfortunately, HK2 is still the only DI-implementation available for Jersey and it doesn't look like that's going to change soon. 🤔

File: dropwizard-jersey/src/main/java/io/dropwizard/jersey/validation/NonEmptyStringParamValueExtractor.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+import org.hibernate.validator.internal.engine.valueextraction.ValueExtractorDescriptor;
+
+import javax.validation.valueextraction.ExtractedValue;
+import javax.validation.valueextraction.UnwrapByDefault;
+import javax.validation.valueextraction.ValueExtractor;
+
+/**
+ * Let's the validator know that when validating a {@link NonEmptyStringParam} to validate the
+ * underlying value. This class is needed, temporarily, while Hibernate is not able to unwrap nested
+ * classes <a href="https://hibernate.atlassian.net/browse/HV-904"/>.
```
[the remark this answers is not in the document]
Comment (joschi) [reply dropwizard-jersey/src/main/java/io/dropwizard/jersey/validation/NonEmptyStringParamValueExtractor.java]: @nickbabcock Given that we could replace `NonEmptyStringParam` by `Optional<@NotEmpty String>` with the latest version of Hibernate Validator, should we simply remove the type?

File: dropwizard-jersey/src/main/java/io/dropwizard/jersey/optional/OptionalIntParamConverterProvider.java [hunk 7]
```diff
@@ -26,14 +26,11 @@
     public static class OptionalIntParamConverter implements ParamConverter<OptionalInt> {
         @Override
         public OptionalInt fromString(final String value) {
-            if (value == null) {
```
[the remark this answers is not in the document]
Comment (joschi) [reply dropwizard-jersey/src/main/java/io/dropwizard/jersey/optional/OptionalIntParamConverterProvider.java]: @nickbabcock Maybe we should remove these types altogether? They can be replaced by an appropriate optional type, for example `OptionalInt`.

**Comment by usulkies** [comment]
> @usulkies Good catch, thanks! Thankfully, it was quite simple to replace the HK2 binders with Jersey binders.
> 
> Unfortunately, HK2 is still the only DI-implementation available for Jersey and it doesn't look like that's going to change soon. 🤔

Yes I know. In my company we're working with Dropwizard 1.3.x and Google Guice DI, and we're currently using [dropwizard-guicier](https://github.com/HubSpot/dropwizard-guicier), which in turn uses [jersey2-guice ](https://github.com/Squarespace/jersey2-guice) to bind Guice with Jersey. I don't know yet if and how we'll be able to move to DW 2.0 when it is released.
I've opened a [PR](https://github.com/Squarespace/jersey2-guice/pull/44) to the latter, but it seems that this project is no longer maintained.
I searched if someone tried to implement Jersey's `InjectionManager` for Guice, but except for this not-recently-touched [issue](https://github.com/jersey/jersey/issues/3692) - found none.
Will keep try and looking for a solution. Please let me know if you find any. Keep the good work!

**Comment by jplock** [comment]
@joschi we should probably do another 1.3.x release with everything being removed in 2.0.0 deprecated, to give people a head’s up. Is it just IntParam and LongParam?

File: docs/source/manual/validation.rst [hunk 8]
```diff
@@ -76,80 +76,46 @@ detailing the validation errors: ``name may not be empty``
...
-
-.. code-block:: java
-
-    @POST
-    // The @NotNull is supposed to mean that the parameter is required but the Max(3) is supposed to
-    // apply to the contained integer. Currently, this code will fail saying that Max can't
-    // be applied on an IntParam
-    public List<Person> createNum(@QueryParam("num") @UnwrapValidatedValue(false)
-                                  @NotNull @Max(3) IntParam num) {
+    public List<Person> createName(@QueryParam("name") Optional<@NotBlank String> num) {
```
Comment (nickbabcock) [inline docs/source/manual/validation.rst]: Code snippet is either out of place, or needs a `.. code-block:: java`
